File: NetDeck/CardCounter.m
//
//  CardCounter.m
//  Net Deck
//
//  Created by Gereon Steffens on 24.11.13.
//  Copyright (c) 2015 Gereon Steffens. All rights reserved.
//

#import "CardCounter.h"
#import "Card.h"

@implementation CardCounter

@synthesize card = _card;

+(CardCounter*) initWithCard:(Card*)card
{
    return [self initWithCard:card andCount:1];
}

+(CardCounter*) initWithCard:(Card*)card andCount:(NSInteger)count
{
    NSAssert(card != nil, @"card is nil");
    CardCounter* cc = [CardCounter new];
    cc->_card = card;
    cc.count = count;
        
    return cc;
}

#pragma mark NSCoding

-(id) initWithCoder:(NSCoder *)decoder
{
    if ((self = [super init]))
    {
        NSString* code = [decoder decodeObjectForKey:@"card"];
        _card = [Card cardByCode:code];
        self.count = [decoder decodeIntegerForKey:@"count"];
    }
    return self;
}

-(void) encodeWithCoder:(NSCoder *)coder
{
    [coder encodeInteger:self.count forKey:@"count"];
    [coder encodeObject:self.card.code forKey:@"card"];
}

@end

File: NetDeck/CardCounter.swift
//
//  CardCounter.swift
//  NetDeck
//
//  Created by Gereon Steffens on 14.11.15.
//  Copyright © 2021 Gereon Steffens. All rights reserved.
//

import Foundation

@objc(CardCounter) class CardCounter: NSObject, NSCoding, NSCopying {
    @objc let card: Card
    var count: Int
    
    private static let nullInstance = CardCounter(card: Card.null(), count: 0)
    
    init(card: Card, count: Int) {
        self.card = card
        self.count = count
    }
    
    static func null() -> CardCounter {
        return nullInstance
    }
    
    var isNull: Bool {
        return self === CardCounter.nullInstance || self.card.isNull
    }

    func displayName(_ mwl: Int) -> String {
        if self.card.type == .identity {
            return self.card.displayName(mwl)
        } else {
            return self.card.displayName(mwl, count: self.count)
        }
    }
    
    // MARK: NSCoding
    convenience required init?(coder aDecoder: NSCoder) {
        let code = aDecoder.decodeObject(forKey: "card") as! String
        let count = aDecoder.decodeInteger(forKey: "count")
        let card = CardManager.cardBy(code, useReplacements: false) ?? Card.null()
        
        self.init(card: card, count: card.isNull ? 0 : count)
    }
    
    func encode(with aCoder: NSCoder) {
        aCoder.encode(self.count, forKey: "count")
        aCoder.encode(self.card.code, forKey: "card")
    }

    // MARK: NSCopying
    func copy(with zone: NSZone?) -> Any {
        return CardCounter(card: self.card, count: self.count)
    }
}
